### ISSUE TITLE
Support `cargo:info=MESSAGE` in `build.rs` stdout

### ISSUE THREAD

**Issue by fredpointzero** [rust-lang/cargo#7037]
**Describe the problem you are trying to solve**
Provide meaningful information to a developer when using a `build.rs` script.
In my case, I provide duration of specific steps to monitor the performance of thoses.

Currently, I use `cargo:warning=MESSAGE`  to provide this output.

This is not fine because it is not an actual warning.

**Describe the solution you'd like**

Supporting  `cargo:info=MESSAGE` to output a string in the cargo's output with info log level.

Example:
(build.rs)
```
fn main() {
     println!("cargo:info=Hello World!");
}
```

Will print in the cargo output if the `build.rs` script is executed:
```
info: Hello World!
```
with the look of an info entry.

**Comment by Fishrock123** [comment]
This would be quite useful.

**Comment by davidkern** [comment]
I was looking at putting up a PR for this, but after looking at the code, I can see there is some more discussion needed first.

Internally, cargo has support for printing "warn", "error" and "note" messages to the shell. For consistency, I suggest this feature should be implemented as "cargo:note=MESSAGE" which would print the cyan note message.  Additionally, similar functionality should be implemented for "error" as well (presumably error would also trigger cargo to abort).

A complication with the existing architecture is that messages are collected from the custom build script and are not written to the console until after the build script has completed.  Would this restriction still support the motivating use cases for this issue?

In my use case, waiting until the end of the build script would not be sufficient. I have a build which requires the presence of a 500MB archive, which is downloaded if not present. I want to display a note to the user that this download is in progress, otherwise the impression is that cargo build has hung, particularly on slow connections.

**Comment by joshtriplett** [comment]
We talked about this in today's @rust-lang/cargo meeting.

A summary of the rough consensus:

- This needs design work, and an RFC; we don't just want to do this via PR.
- Info messages should not show up by default, only if you pass `-v` or similar. We're concerned about this being misused if visible by default. The equivalent in npm, for instance, is sometimes used for advertisements.
- We'd be interested in a more narrowly scoped mechanism for `cargo:progress=` messages, which would *only* carry numeric progress information (e.g. percentage done); Cargo's own progress could then show something like `cratename(build x%)`, and that *could* be shown by default. Such a mechanism would need to allow for multiple progress indicators from the same crate, because this would almost certainly come primarily from helper crates rather than from build scripts directly. (For instance, the `cc` or `cmake` crates could give progress indicators for their builds, or a common download helper crate could provide progress indicators for its downloads.) We don't want such a mechanism to introduce too much complexity, but a simple tag+number system might work.

**Comment by natalia-asteria** [comment]
> Info messages should not show up by default, only if you pass -v or similar. We're concerned about this being misused if visible by default. The equivalent in npm, for instance, is sometimes used for advertisements.

Yeah, that would be a problem, but I think this might be useful too for . We can perhaps allow logs from certain crates with configuration or/and crates that are locally hosted (rather than fetched from Crates.io)?

**Comment by codecnotsupported** [comment]
If you want a workaround you can abuse escape codes to **overwrite** the line with something new.

```
macro_rules! p {
    ($($tokens: tt)*) => {
        println!("cargo:warning=\r\x1b[32;1m   {}", format!($($tokens)*))
    }
}
```
```
p!("Compiling XYZ");
```

**Comment by epage** [comment]
This is also blocked on #11461 being resolved..

**Comment by weihanglo** [comment]
I overlooked at that time. <https://github.com/rust-lang/cargo/issues/11593> seems kinda a duplicate of this?

**Comment by epage** [comment]
Isn't #11593 about non-cap-lint build script warnings while this is about info messages?

**Comment by newcomertv** [comment]
Hey, what's the status on this?

There are legitimate use cases for printing general build info like when using an environment variable that will be overridden with a default value if not found.

Anything I can help with?

**Comment by epage** [comment]
Within cargo's current model, that sounds like something that would be gated behind verbose  output. iirc build script output is shown on verbose (or is it very verbose) output. What additional value would `info` then provide over printing it yourself?

**Comment by tklajnscek** [comment]
Hey, I stumbled back into this after a few years to find it still open :)

I think the key distinction here is that you want to show some kind of potentially important message, but not the entire output of the build script.

Our use case is - we have `cargo.toml` metadata field that can switch between building CMake dependencies locally vs downloading them prebuilt. It would be super nice if I could output:
`Building CMake dependencies!`
or 
`Downloading pre-built dependencies!` 
within the usual progress output cargo prints already such as:
```
==================== Compiling executable ====================
   Compiling bla v0.0.23
   Compiling xyzv0.0.23 
```

**Comment by epage** [comment]
Not that the "currrent operation" ("Compiling") would be unlikely to be overriden because that reflects a concurrent operation and trying to analyze what should be shown would be tricky. This just leaves the sticky result ("Compiled"). I can understand wanting to show nuance for very different build script operations but we also have a lot of people not thrilled with how much we output (on top of the abuse seen in other ecosystemsb) and we're looking at redoing our output in #8889. What we'd want from an "info status" after that may be very different

**Comment by ic3man5** [comment]
This would be useful to have. I've never been a fan of the syntax though. Is there a reason we can't just use `log::{info, debug, warn}`?

**Comment by atlasoft** [comment]
I think this is open due the existence of [build-print](https://docs.rs/build-print/latest/build_print/) crate. It is quite useful, but it's still cumbersome that stdout is buffered, so if the output of the build.rs script is too long, the actual compilation sits in the middle and sometimes the lines break

**Comment by epage** [comment]
@atlasoft as a heads up, we're considering stripping ANSI escape codes emitted from build scripts, see #15829.

**Comment by tgross35** [comment]
Some thoughts: I agree that this should be off by default, but think that having this enabled with `-v` may be more coarse grained than ideal. For some crates `-v` enables a _lot_ of output (lots of `check-cfg` leading to long rustc invocation) and it seems that helpful info may get lost in noise that isn't otherwise needed.

An idea popped into my head that I've started to like a lot: what if verbose configuration allowed a syntax similar to env_logger? Something like the following:

```sh
# Enable printing `cargo::info=MESSAGE` for all crates
cargo build -vcrates=info

# Can be set via env
CARGO_TERM_VERBOSE=crates=info cargo build

# Enable info logging for a specific crate or set of crates
CARGO_TERM_VERBOSE=my-foo=info cargo build
CARGO_TERM_VERBOSE=my-foo,my-bar=info cargo build
CARGO_TERM_VERBOSE=my-.*=info cargo build

# "Trace" logging shows everything that `my_crate`'s build.rs prints
CARGO_TERM_VERBOSE=my_crate=trace cargo build

# Maybe also allow this to configure Cargo's output? Since `-v` is used
CARGO_TERM_VERBOSE=cargo=info cargo build # same as -v
CARGO_TERM_VERBOSE=cargo=debug cargo build # same as -vv
CARGO_TERM_VERBOSE=cargo=trace cargo build # same as -vvv

# Combinations
CARGO_TERM_VERBOSE=cargo=info,crates=debug,my-special-crate=trace cargo build
```

I think this would do really a nice job of solving some annoyances that I've definitely run into:

* You can enable build script info output without increasing Cargo's verbosity
* When writing a build script, you can use `CARGO_TERM_VERBOSE=my_crate=trace` to see output of `println!`. No need to wrap in `cargo::info=` or `cargo::warning=`, which is a convenience.
* You can enable output for only specific crates. This is great when you need to debug a crate that is using `cc` because it would allow setting `CC_ENABLE_DEBUG_OUTPUT ` and seeing the output for a single crate. Currently the only thing that can be done is to set that env and using `cargo ... -vv`, which will print output for _all_ `-sys` crates and can be incredibly noisy.

**Comment by lexi-the-cute** [comment]
Note: If y'all require me to open a separate issue for the below suggestion
instead of tacking on to this existing issue, let me know, and the next time I
see a notification, I will correct this and create a new issue instead.



I very much would like to have:

* `println!("cargo::trace={}", "hello, trace...");`
* `println!("cargo::debug={}", "hello, debug...");`
* `println!("cargo::info={}", "hello, info...");`
* `println!("cargo::warning={}", "hello, warn...");`
* `println!("cargo::error={}", "hello, error...");`
* `println!("cargo::fatal={}", "hello, fatal, instead of trace. see below...");`

  Where trace is for well, dumping the stack trace with some useful developer
memo, debug is for lots of useful debugging info, but most likely will not
need to be displayed by default, info would be for information which the crate
author believes might be useful for the user to know about, but is not a
warning, nor an error, and then the already existing warning and error lines.

  I can see a case where trace is not needed and to only have the stack trace
printed when error is called, or if similar behavior is desired, but not
changing the existing behavior of `cargo::error`, then `cargo::fatal` could be
considered be considered if we desire any form of stack trace at all.

**Comment by epage** [comment]
We have `cargo::warning` and `cargo::error` today.  I'm not sure what adding a `cargo::fatal` would add.

Might be worth considering not just what categories we should support but how users are meant to interact with them, including
- when or how would the user see them that makes them different
- what, if any, side effects they should have (e.g. `cargo::error` causes the build script to fail)
